### PR TITLE
[cilium] Add safe-agent-updater

### PR DESCRIPTION
**Pull request by AndreyPavlovFlant**
## Description

To ensure a smoother and more seamless update process for cilium-agent pods, we will take the following steps:
- We will set the parameter `updateStrategy.type: OnDelete` for the main DaemonSet
- We will create a service DaemonSet that:
  - Loads a new agent image onto the node
  - If any changes are made to the main DaemonSet, it reboots its pods (this is necessary since above we disabled automatic reboots of pods) 

## Why do we need it, and what problem does it solve?

When updating the cilium-agent version or modifying its image, you may encounter the following situation:
- the “old” pod has already been deleted from the node, and the new pod is taking a long time to start due to problems with loading the new image. 
  Depending on the functionality used, this may lead to problems with the network.

## What is the expected result?

Updating and restarting cilium-agent pods is seamless and smoother.

## Checklist
- [x] The code is covered by unit tests.
- [x] e2e tests passed.
- [ ] Documentation updated according to the changes.
- [x] Changes were tested in the Kubernetes cluster manually.

## Changelog entries

```changes
section: cilium
type: fix
summary: Adding safe-agent-updater
impact: Cilium-agent pods will be restarted.
impact_level: default
```

### CODE REVIEW

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/openstack'...

**Comment by AndreyPavlovFlant** [comment]
/e2e/abort openstack;Standard;containerd;1.25 7758934714 7758934714-1-con-1-25 /sys/deckhouse-oss/install:pr7333

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by z9r5** [comment]
/e2e/abort openstack;Standard;containerd;1.25 7758934714 7758934714-1-con-1-25 /sys/deckhouse-oss/install:pr7333

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @z9r5. I've recognized your '/e2e/abort' command and started the workflow...

File: modules/021-cni-cilium/templates/safe-updater/rbac-for-us.yaml [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+apiVersion: rbac.authorization.k8s.io/v1
+metadata:
+  name: d8:{{ .Chart.Name }}:safe-updater
+  {{ include "helm_lib_module_labels" (list . (dict "app" "safe-updater")) | nindent 2 }}
+rules:
+- apiGroups:
+    - ""
+  resources:
+    - namespaces
+    - nodes
```
Comment (apolovov) [inline modules/021-cni-cilium/templates/safe-updater/rbac-for-us.yaml:18]: we don't need it

File: modules/021-cni-cilium/templates/safe-updater/rbac-for-us.yaml [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+  resources:
+    - pods
+  verbs:
+    - get
+    - list
+    - watch
+    - create
+    - update
+    - delete
+    - patch
```
Comment (apolovov) [inline modules/021-cni-cilium/templates/safe-updater/rbac-for-us.yaml:34]:
```suggestion
    - get
    - list
    - delete
```

File: modules/021-cni-cilium/templates/safe-updater/rbac-for-us.yaml [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+- apiGroups:
+  - apps
+  resources:
+  - daemonsets
+  verbs:
+  - get
+  - list
+  - watch
```
Comment (apolovov) [inline modules/021-cni-cilium/templates/safe-updater/rbac-for-us.yaml:42]:
```suggestion
```

File: modules/021-cni-cilium/templates/safe-updater/rbac-for-us.yaml [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+- apiGroups:
+    - events.k8s.io
+  resources:
+    - events
+  verbs:
+    - create
+    - update
+    - list
+    - delete
+    - patch
```
Comment (apolovov) [inline modules/021-cni-cilium/templates/safe-updater/rbac-for-us.yaml:59]:
```suggestion
```

File: modules/021-cni-cilium/images/safe-updater/main.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+
+	CiliumAgentDS, err := kubeClient.AppsV1().DaemonSets(ciliumNs).Get(
+		context.TODO(),
+		"agent",
+		metav1.GetOptions{},
+	)
+	if err != nil {
+		log.Error(err)
+	}
+	CiliumAgentDSGenerationChecksum := CiliumAgentDS.Spec.Template.Annotations[generationChecksumAnnotations]
```
[the remark this answers is not in the document]
Comment (apolovov) [reply modules/021-cni-cilium/images/safe-updater/main.go:60]: Throw an error if there isn't an annotation.

**Review by apolovov — CHANGES_REQUESTED** [review]

File: modules/021-cni-cilium/images/safe-agent-updater/main.go [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+			if err != nil {
+				log.Errorf("[SafeAgentUpdater] Failed to list pods on same node. Error: %v", err)
+			}
+			log.Infof(
+				"[SafeAgentUpdater] Count of agents running on node %s is %v",
+				nodeName,
+				len(ciliumAgentPodsOnSameNode.Items),
+			)
+
+			if len(ciliumAgentPodsOnSameNode.Items) == 1 {
```
Comment (apolovov) [inline modules/021-cni-cilium/images/safe-agent-updater/main.go:146]: and kind of ciliumAgentPodsOnSameNode.Items[0].deletionTimestamp == null

**Review by apolovov — CHANGES_REQUESTED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/static'...

File: modules/021-cni-cilium/images/safe-agent-updater/main.go [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+	config, _ := rest.InClusterConfig()
+	kubeClient, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		log.Fatalf("[SafeAgentUpdater] Failed to init kubeClient. Error: %v", err)
+	}
+	nodeName := os.Getenv("NODE_NAME")
+	if len(nodeName) == 0 {
+		log.Fatalf("[SafeAgentUpdater] Failed to get env NODE_NAME.")
+	}
+	currentPodName, isCurrentGenEqDesiredGen, err := checkGeneration(kubeClient, nodeName)
```
Comment (apolovov) [inline modules/021-cni-cilium/images/safe-agent-updater/main.go:49]:
```suggestion
	currentAgentPodName, isCurrentAgentPodGenerationDesired, err := checkAgentPodGeneration(kubeClient, nodeName)
```

curren
i don't like condition in varname. isAgentPodGenerationDesired.

File: modules/021-cni-cilium/images/safe-agent-updater/main.go [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+			"[SafeAgentUpdater] Failed to get DaemonSets %s/agent. Error: %v",
+			ciliumNS,
+			err,
+		)
+	}
+
+	desiredAgentGeneration := ciliumAgentDS.Spec.Template.Annotations[generationAnnotation]
+	if len(desiredAgentGeneration) == 0 {
+		return "", false, fmt.Errorf(
+			"[SafeAgentUpdater] DaemonSets %s/agent doesn't have annotations %s",
```
Comment (apolovov) [inline modules/021-cni-cilium/images/safe-agent-updater/main.go:83]:
```suggestion
			"[SafeAgentUpdater] DaemonSets %s/agent doesn't have annotation %s",
```

File: modules/021-cni-cilium/images/safe-agent-updater/main.go [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+		currentPod.Name,
+	)
+	currentAgentGeneration := currentPod.Annotations[generationAnnotation]
+	log.Infof(
+		"[SafeAgentUpdater] Generation of pod %s is %s",
+		currentPod.Name,
+		currentAgentGeneration,
+	)
+
+	switch {
```
Comment (apolovov) [inline modules/021-cni-cilium/images/safe-agent-updater/main.go:137]:
```suggestion
if desiredAgentGeneration == currentAgentGeneration {
  log
  return  currentPod.Name, true, nil
} else {
  return  currentPod.Name, false, nil
}
```

File: modules/021-cni-cilium/images/safe-agent-updater/main.go [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+		"[SafeAgentUpdater] Pod %s/%s deleted",
+		ciliumNS,
+		podName,
+	)
+	return nil
+}
+
+func waitUntilNewPodCreatedAndBecomeReady(kubeClient kubernetes.Interface, nodeName string, scanIterations int) error {
+	var newPodName string
+	for i := 1; i <= scanIterations; i++ {
```
Comment (apolovov) [inline modules/021-cni-cilium/images/safe-agent-updater/main.go:179]:
```suggestion
	for i := 0; i < scanIterations; i++ {
```
classic

File: modules/021-cni-cilium/images/safe-agent-updater/main.go (hunk 9 above)
Comment (apolovov) [inline modules/021-cni-cilium/images/safe-agent-updater/main.go:137]: if else return

File: modules/021-cni-cilium/images/safe-agent-updater/werf.inc.yaml [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+  install:
+  - cd /src
+  - go mod download
+  setup:
+  - export GOOS=linux GOARCH=amd64
+  - export CGO_ENABLED=0
+  - cd /src
+  - go build -ldflags="-s -w" -o safe-agent-updater .
+  - chown 64535:64535 /src/safe-agent-updater
+  - chmod 0700 /src/safe-agent-updater
```
Comment (apolovov) [inline modules/021-cni-cilium/images/safe-agent-updater/werf.inc.yaml:29]:
```suggestion
  - chmod +x /src/safe-agent-updater
```

**Review by apolovov — CHANGES_REQUESTED** [review]

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/static'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/openstack'...

**Comment by AndreyPavlovFlant** [comment]
/e2e/abort openstack;Standard;containerd;1.25 7818559401 7818559401-1-con-1-25 /sys/deckhouse-oss/install:pr7333 debian@188.124.39.71

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've recognized your '/e2e/abort' command and started the workflow...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/openstack'...

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/static'...

**Review by apolovov — APPROVED** [review]

File: modules/021-cni-cilium/images/safe-agent-updater/main.go [hunk 12]
```diff
@@ -0,0 +1,245 @@
...
+	)
+
+	if desiredAgentGeneration == currentAgentGeneration {
+		log.Infof(
+			"[SafeAgentUpdater] Desired agent generation(%s) and current(%s) are same. Nothing to do.",
+			desiredAgentGeneration,
+			currentAgentGeneration,
+		)
+		return currentPod.Name, true, nil
+	} else {
```
Comment (RomanenkoDenys) [inline modules/021-cni-cilium/images/safe-agent-updater/main.go:144]: why we need else case here ? In the if we return so we do not need else.

File: modules/021-cni-cilium/templates/agent/daemonset.yaml [hunk 13]
```diff
@@ -458,3 +461,8 @@ spec:
                 path: server.crt
               - key: tls.key
                 path: server.key
+{{- end  }}
+
+{{ $agent_daemonset_generation := include "agent_daemonset_template" . | sha256sum }}
+
+{{- include "agent_daemonset_template" . | replace "%%DS_GENERATION%%" $agent_daemonset_generation }}
```
Comment (RomanenkoDenys) [inline modules/021-cni-cilium/templates/agent/daemonset.yaml:468]: please do not use replace, use template variables

File: modules/021-cni-cilium/templates/safe-agent-updater/daemonset.yaml [hunk 14]
```diff
@@ -0,0 +1,85 @@
...
+              apiVersion: v1
+              fieldPath: spec.nodeName
+        resources:
+          requests:
+            {{- include "helm_lib_module_ephemeral_storage_only_logs" . | nindent 12 }}
+            {{- include "safe_agent_updater_resources" . | nindent 12 }}
+      containers:
+      - name: pause
+        {{- include "helm_lib_module_container_security_context_read_only_root_filesystem" . | nindent 8 }}
+        image: {{ include "helm_lib_module_common_image" (list . "pause") }}
```
Comment (apolovov) [inline modules/021-cni-cilium/templates/safe-agent-updater/daemonset.yaml:77]: There must be cilium-agent image with pause binary like in [d8-control-plane-manager](https://github.com/deckhouse/deckhouse/blob/main/modules/040-control-plane-manager/images/kube-apiserver/werf.inc.yaml#L9).

File: modules/021-cni-cilium/templates/agent/daemonset.yaml (hunk 13 above)
Comment (apolovov) [reply modules/021-cni-cilium/templates/agent/daemonset.yaml:468]: @AndreyPavlovFlant need to add $context var and replace all dots to $context in template.

**Comment by deckhouse-BOaTswain** [comment]
Aye, aye, @AndreyPavlovFlant. I've started the workflow for label 'e2e/run/static'...

**Comment by AndreyPavlovFlant** [comment]
/e2e/abort static;Static;containerd;1.25 7901720196 7901720196-1-con-1-25 /sys/deckhouse-oss/install:pr7333

File: modules/021-cni-cilium/templates/agent/daemonset.yaml [hunk 15]
```diff
@@ -13,22 +13,27 @@ spec:
   {{- end }}
 {{- end }}
 ---
+{{- define "agent_daemonset_template" }}
+  {{- $context := index . 0 }}
+  {{- $agent_daemonset_generation := index . 1 | default "undefined"}}
```
Comment (apolovov) [inline modules/021-cni-cilium/templates/agent/daemonset.yaml:18]:
```suggestion
  {{- $agent_daemonset_generation := index . 1 }}
```

**Review by RomanenkoDenys — APPROVED** [review]

**Review by name212 — APPROVED** [review]

**Pull request merged**